pkg/cache/server/handler.go

```diff
@@ -0,0 +1,112 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package server
+
+import (
+	"fmt"
+	"net/http"
+	"regexp"
+	"strings"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apiserver/pkg/endpoints/handlers/responsewriters"
+	"k8s.io/apiserver/pkg/endpoints/request"
+)
+
+var (
+	shardNameRegExp = regexp.MustCompile(`^([a-z]([a-z0-9-]{0,61}[a-z0-9])?:)*[a-z]([a-z0-9-]{0,61}[a-z0-9])?$`)
+
+	errorScheme = runtime.NewScheme()
+	errorCodecs = serializer.NewCodecFactory(errorScheme)
+)
+
+func init() {
+	errorScheme.AddUnversionedTypes(metav1.Unversioned,
+		&metav1.Status{},
+	)
+}
+
+// WithShardScope reads a shard name from the URL path and puts it into the context.
+// It also trims "/shards/" prefix from the URL.
+// If the path doesn't contain the shard name then a default "cache" name is assigned.
+//
+// For example:
+//
+// /shards/*/clusters/*/apis/apis.kcp.dev/v1alpha1/apiexports'
+//
+// /shards/sapphire/clusters/system:sapphire/apis/apis.kcp.dev/v1alpha1/apiexports'
+//
+// /shards/amber/clusters/system:amber/apis/apis.kcp.dev/v1alpha1/apiexports'
+func WithShardScope(handler http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		var shardName string
+		if path := req.URL.Path; strings.HasPrefix(path, "/shards/") {
+			path = strings.TrimPrefix(path, "/shards/")
+
+			i := strings.Index(path, "/")
+			if i == -1 {
+				responsewriters.ErrorNegotiated(
+					apierrors.NewBadRequest(fmt.Sprintf("unable to parse cluster: no `/` found in path %s", path)),
+					errorCodecs, schema.GroupVersion{},
+					w, req)
+				return
+			}
+			shardName, path = path[:i], path[i:]
+			req.URL.Path = path
+			for i := 0; i < 2 && len(req.URL.RawPath) > 1; i++ {
+				slash := strings.Index(req.URL.RawPath[1:], "/")
+				if slash == -1 {
+					responsewriters.ErrorNegotiated(
+						apierrors.NewInternalError(fmt.Errorf("unable to parse shard when shortening raw path, have shardName=%q, rawPath=%q", shardName, req.URL.RawPath)),
+						errorCodecs, schema.GroupVersion{},
+						w, req)
+					return
+				}
+				req.URL.RawPath = req.URL.RawPath[slash:]
+			}
+		}
+
+		var shard request.Shard
+		switch {
+		case shardName == "*":
+			shard.Wildcard = true
+			shard.Name = ""
+		case len(shardName) == 0:
+			// this is really important, usually it means that the request is from the loopback client.
+			// we set the name here because we don't store a shard name in an object.
+			// that means a wildcard list request returns objects without associated shard.
+			// thus updating an object is problematic because we won't be able to find the original object without a shard name.
```

Comment:
cannot follow this case

Reply:
we need it to make kube internal clients shard aware, e.g. the `EstablishingController` otherwise they will just break.

Reply:
the same applies to kcp, it is just hidden by `multiClusterClientConfigRoundTripper` :)

Reply:
which other controllers? Only those from apiextensions?

This needs a high level comment, above "this is really important".

What does "shard = cache" mean? It's a fake-shard for the CRDs?

Reply:
> which other controllers? Only those from apiextensions?

yes, apiextention controllers.

> What does "shard = cache" mean? It's a fake-shard for the CRDs?

yes, open to suggestions. Maybe we should change it to some since we prefix all resources with a '/cache/group/resource` already.

Reply:
maybe `system:something` ?

Reply:
so it does not conflict with shard names